providers/multi/provider.go

```diff
@@ -60,21 +68,96 @@ func New(opts Options) *Provider {
 	if p.opts.Separator == "" {
 		p.opts.Separator = "#"
 	}
+	if p.opts.ChannelSize <= 0 {
+		p.opts.ChannelSize = 10
+	}
 
 	p.log = log.Log.WithName("multi-provider")
 
+	p.indexers = make([]index, 0)
 	p.providers = make(map[string]multicluster.Provider)
-	p.providerCancel = make(map[string]context.CancelFunc)
+	p.providersCancel = make(map[string]context.CancelFunc)
 
 	return p
 }
 
-// SetManager sets the manager for the provider.
-func (p *Provider) SetManager(mgr mctrl.Manager) {
-	if p.mgr != nil {
-		p.log.Error(nil, "manager already set, overwriting")
+// Start runs the provider. It runs all providers that implement
+// multicluster.ProviderRunnable, even those added after Start() has
+// been called.
+func (p *Provider) Start(ctx context.Context, aware multicluster.Aware) error {
+	p.log.Info("starting multi provider")
+
+	p.lock.Lock()
+	p.prefixCh = make(chan string, p.opts.ChannelSize)
+	prefixes := maps.Keys(p.providers)
+	p.lock.Unlock()
+
+	for prefix := range prefixes {
+		p.startProvider(ctx, prefix, aware)
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case prefix := <-p.prefixCh:
+			p.startProvider(ctx, prefix, aware)
+		}
 	}
-	p.mgr = mgr
+}
+
+func (p *Provider) startProvider(ctx context.Context, prefix string, aware multicluster.Aware) {
+	p.log.Info("starting provider", "prefix", prefix)
+
+	p.lock.RLock()
+	provider, ok := p.providers[prefix]
+	p.lock.RUnlock()
+	if !ok {
+		p.log.Error(nil, "provider not found", "prefix", prefix)
+		return
+	}
+
+	runnable, ok := provider.(multicluster.ProviderRunnable)
+	if !ok {
+		p.log.Info("provider is not runnable, not starting", "prefix", prefix)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	wrappedAware := &wrappedAware{
+		Aware:  aware,
+		prefix: prefix,
+		sep:    p.opts.Separator,
+	}
+
+	p.lock.Lock()
+	if _, ok := p.providersCancel[prefix]; ok {
+		// This is a failsafe. It should never happen but on the off
+		// change that it somehow does the provider shouldn't be started
+		// twice.
+		cancel()
+		p.log.Error(nil, "provider already started, not starting again", "prefix", prefix)
+		p.lock.Unlock()
+		return
+	}
+	p.providersCancel[prefix] = cancel
+	p.lock.Unlock()
+
+	go func() {
+		defer p.RemoveProvider(prefix)
+		if err := runnable.Start(ctx, wrappedAware); err != nil {
```

Comment:
Do we still need this?
wondering if WrapperAware will not implement the Runnable interface, and the manager could start them too?